azkaban-common/src/main/java/azkaban/executor/container/ContainerImplUtils.java

```diff
@@ -82,4 +93,80 @@ public static int getFlowNameHashValMapping(ExecutableFlow flow) {
+      if (!proxyUserFromFlowProp.isEmpty()) {
+        proxyUsers.add(proxyUserFromFlowProp);
+      }
+    }
+    // DFS Walk of the Graph to find all the Proxy Users.
+    populateProxyUsersForFlow(flow, flowObj, project, projectManager, proxyUsers);
+    return proxyUsers;
+  }
+
+  public static void populateProxyUsersForFlow(final ExecutableNode node, final Flow flowObj,
+      final Project project, final ProjectManager projectManager, Set<String> proxyUsers) {
+    if (node instanceof ExecutableFlowBase) {
+      final ExecutableFlowBase base = (ExecutableFlowBase) node;
+      for (ExecutableNode subNode : base.getExecutableNodes()) {
+        populateProxyUsersForFlow(subNode, flowObj, project, projectManager, proxyUsers);
+      }
+    } else {
+      // If a node is disabled, we don't need to initialize its jobType container image when
+      // creating a pod.
+      if (node.getStatus() != Status.DISABLED) {
+        Props currentNodeProps = projectManager.getProperties(project, flowObj,
+            node.getId(), node.getJobSource());
+        // Get the node level property for proxy user.
+        String userToProxyFromNode = currentNodeProps.getString(USER_TO_PROXY, "");
+        // Get the node level override by user from the UI for proxy user.
+        Props currentNodeJobProps = projectManager.getJobOverrideProperty(project, flowObj,
+            node.getId(), node.getJobSource());
+        String userToProxyFromJobNode = currentNodeJobProps.getString(USER_TO_PROXY, "");
+        if (!userToProxyFromJobNode.isEmpty()) {
+          proxyUsers.add(userToProxyFromJobNode);
+        } else if (!userToProxyFromNode.isEmpty())
+          proxyUsers.add(userToProxyFromNode);
+      }
+    }
+  }
+
+/* Extract the proxy users needed from  PREFETCH_JOBTYPE_PROXY_USER_MAP
+   This method is being introduced to be able to assign a specific proxy user that will require
+   custom credentials for a given job type. This allows the jobtype to perform specific checks
+   without requiring azkaban executor's credentials to do this, once we enforce POLP defined in
+   https://github.com/azkaban/azkaban/pull/3216
+   This will parse the jobTypePrefetchUserMap of the format : "jobtype1,jobtype1_proxyuser;
+   jobtype2,jobtype2_proxyuser" and add the proxy user for a given job if that jobtype is
+   present in the flow.
+ */
+
+  public static Set<String> getJobTypeUsersForFlow(String jobTypePrefetchUserMap,
+      TreeSet<String> jobTypes) {
+    Set<String> jobTypeProxyUserSet = new HashSet<>();
+    StringTokenizer st = new StringTokenizer(jobTypePrefetchUserMap, ";");
+    while (st.hasMoreTokens()) {
+      StringTokenizer stInner = new StringTokenizer(st.nextToken(), ",");
+      String jobType = stInner.nextToken();
+      String jobTypeUser = stInner.nextToken();
+      if (jobTypes.contains(jobType)) {
+        jobTypeProxyUserSet.add(jobTypeUser);
+      }
+    }
+    return jobTypeProxyUserSet;
+  }
```

Comment:
Since the `jobTypePrefetchUserMap` will not change from one execution to the next, to avoid parsing this string everytime can we please do it just once in the constructor of KubernetesContainerizedImpl.java? The result will be a map of <jobtype1, jobtype1_proxyuser> which can be iterated later to populate jobTypeProxyUserSet. That should be much faster. 
Also, please trim PREFETCH_JOBTYPE_PROXY_USER_MAP when parsing to account for typos.

Reply:
I'll need the jobtypes set to make this set in the constructor. Which is a DAG walk. If i do this in the constructor. Its one extra DAG walk. Not sure if there's any other way to achieve this.

Reply:
resolved this by creating a new function